include/realtime_tools/realtime_box.h

```diff
@@ -26,50 +26,402 @@
+    func(value_);
+  }
+
+  /**
+   * @brief get the content and wait until the mutex could be locked (RealtimeBox behaviour)
+   * @return copy of the value
+   */
+  template <typename U = T>
+  [[nodiscard]] typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, U> get() const
+  {
+    std::lock_guard<mutex_t> guard(lock_);
+    return value_;
+  }
+  /**
+   * @brief get the content and wait until the mutex could be locked
+   * @note same signature as in the existing RealtimeBox<T>
+   */
+  template <typename U = T>
+  typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, void> get(T & in) const
+  {
+    std::lock_guard<mutex_t> guard(lock_);
+    // cppcheck-suppress missingReturn
+    in = value_;
+  }
+  /**
+   * @brief access the content (r) and wait until the mutex could be locked
+   * @note only safe way to access pointer type content (r)
+   * @note same signature as in the existing RealtimeBox<T>
+   */
+  void get(const std::function<void(const T &)> & func)
+  {
+    std::lock_guard<mutex_t> guard(lock_);
+    func(value_);
+  }
+
+  /**
+   * @brief provide a custom assignment operator for easier usage
+   * @note only to be used from non-RT!
+   */
+  template <typename U = T>
+  typename std::enable_if_t<!is_ptr_or_smart_ptr<U>, void> operator=(const T & value)
+  {
+    set(value);
+  }
 
-  Strongly suggested that you use an std::shared_ptr in this box to
-  guarantee realtime safety.
+  /**
+   * @brief provide a custom conversion operator
+   * @note Can only be used from non-RT!
+   */
+  template <typename U = T, typename = typename std::enable_if_t<!is_ptr_or_smart_ptr<U>>>
+  [[nodiscard]] operator T() const
+  {
+    // Only makes sense with the getNonRT method otherwise we would return an std::optional
+    return get();
+  }
+  /**
+   * @brief provide a custom conversion operator
+   * @note Can be used from non-RT and RT contexts
+   */
+  template <typename U = T, typename = typename std::enable_if_t<!is_ptr_or_smart_ptr<U>>>
+  [[nodiscard]] operator std::optional<T>() const
+  {
+    return tryGet();
+  }
+
+  // In case one wants to actually use a pointer
+  // in this implementation we allow accessing the lock directly.
+  // Note: Be careful with lock.unlock().
+  // It may only be called from the thread that locked the mutex!
+  [[nodiscard]] const mutex_t & getMutex() const { return lock_; }
+  [[nodiscard]] mutex_t & getMutex() { return lock_; }
 
- */
-template <class T>
-class RealtimeBox
+private:
+  T value_;
+  mutable mutex_t lock_;
+};
+
+/**
+ * @brief Specialisation for pointer types.
+ * WHY is this specialised. We do not want to break compatibility but show a deprecation note
+ * for get/set etc. methods if used with a pointer type
+ * They are unsafe to use and should therefore be replace with their correspondents that take an std::function for accessing
+ * the value behind the pointer
+*/
+template <class T, typename mutex_type>
+class RealtimeBox<T, mutex_type, true>
```

Comment:
@firesurfer In this overlayed class, do we need to have a copy of most of the methods similar to the previous one? or is there a way to reuse them?

Reply:
@saikishor From the "base" implementation:

> // Provide a specialisation for non pointer types
>  // NOTE: When migrating to a safe access only version just remove the specialisation for pointer
>  // and let this be the only version!

My argument would be that in the longterm the pointer specialisation should be removed. I think the only save way to access the values behind pointers is to use the "get/set(std::function) overloads). 
By keeping the code duplicated we can simplify this process. 

My suggestion is to keep this version for at least one release and remove the pointer specialisation in the next one.

Reply:
Hmm! @bmagyar @christophfroehlich any strong opinion on this one?